### PR TITLE
Refactor graph-related codes

### PR DESCRIPTION
**Pull request by hyeok9855**
This PR:
1. resolves errors caused by incorrect device assignment when using GPUs.
2. includes several refactorings in `graph_building.py` and related areas throughout the repository.
3. enables the usage of epsilon-greedy (as well as temperature and sf-bias). 

Note that the performance of ring generation remains the same as before.

### CODE REVIEW

File: src/gfn/actions.py [hunk 1]
```diff
@@ -45,14 +45,18 @@ def batch_shape(self) -> tuple[int, ...]:
         return tuple(self.tensor.shape)[: -len(self.action_shape)]
 
     @classmethod
-    def make_dummy_actions(cls, batch_shape: tuple[int, ...]) -> Actions:
+    def make_dummy_actions(
+        cls, batch_shape: tuple[int, ...], device: torch.device
```
Comment (josephdviviano) [inline src/gfn/actions.py:49]: I feel a bit stupid reading this. How does the `device` actually get used?

File: src/gfn/actions.py [hunk 2]
```diff
@@ -45,14 +45,18 @@ def batch_shape(self) -> tuple[int, ...]:
...
+    ) -> Actions:
         """Creates an Actions object of dummy actions with the given batch shape."""
         action_ndim = len(cls.action_shape)
         tensor = cls.dummy_action.repeat(*batch_shape, *((1,) * action_ndim))
         return cls(tensor)
 
     @classmethod
-    def make_exit_actions(cls, batch_shape: tuple[int, ...]) -> Actions:
+    def make_exit_actions(
+        cls, batch_shape: tuple[int, ...], device: torch.device
```
Comment (josephdviviano) [inline src/gfn/actions.py:58]: Ditto.

File: src/gfn/actions.py [hunk 3]
```diff
@@ -126,7 +130,9 @@ def extend_with_dummy_actions(self, required_first_dim: int) -> None:
             if self.batch_shape[0] >= required_first_dim:
                 return
             n = required_first_dim - self.batch_shape[0]
-            dummy_actions = self.__class__.make_dummy_actions((n, self.batch_shape[1]))
+            dummy_actions = self.__class__.make_dummy_actions(
+                (n, self.batch_shape[1]), device=self.device
```
Comment (josephdviviano) [inline src/gfn/actions.py:134]: `self.device` is referencing the `self.tensor.device` - but AFACT the `tensor` does not have a device explicitly set anywhere.

File: src/gfn/actions.py [hunk 4]
```diff
@@ -264,16 +270,20 @@ def edge_index(self) -> torch.Tensor:
...
 
     @classmethod
-    def make_dummy_actions(cls, batch_shape: tuple[int]) -> GraphActions:
+    def make_dummy_actions(
+        cls, batch_shape: tuple[int], device: torch.device
+    ) -> GraphActions:
         """Creates a GraphActions object of dummy actions with the given batch shape."""
         # TODO: make default dtype int32
-        tensor = torch.zeros(batch_shape + (4,), dtype=torch.int64)
+        tensor = torch.zeros(batch_shape + (4,), dtype=torch.int64, device=device)
```
Comment (josephdviviano) [inline src/gfn/actions.py:278]: ok, you do it here -- is it just missing elsewhere?

File: src/gfn/gym/box.py [hunk 5]
```diff
@@ -40,9 +40,12 @@ def __init__(
             exit_action=exit_action,
         )
 
-    def make_random_states_tensor(self, batch_shape: Tuple[int, ...]) -> torch.Tensor:
+    @staticmethod
+    def make_random_states_tensor(
+        batch_shape: Tuple[int, ...], device: torch.device
```
Comment (josephdviviano) [inline src/gfn/gym/box.py:45]: can we draw this information from `self.device` (i.e., whatever device the env is initialized with) by default?

File: src/gfn/gym/discrete_ebm.py [hunk 6]
```diff
@@ -108,14 +108,12 @@ def update_masks(self, states: DiscreteStates) -> None:
         states.backward_masks[..., : self.ndim] = states.tensor == 0
         states.backward_masks[..., self.ndim : 2 * self.ndim] = states.tensor == 1
 
-    def make_random_states_tensor(self, batch_shape: Tuple) -> torch.Tensor:
+    def make_random_states_tensor(
+        self, batch_shape: Tuple, device: torch.device
```
Comment (josephdviviano) [inline src/gfn/gym/discrete_ebm.py:112]: can we draw this information from `self.device` (i.e., whatever device the env is initialized with) by default?

File: src/gfn/gym/graph_building.py [hunk 7]
```diff
@@ -10,35 +11,7 @@
...
-        # Lower triangle.
-        i_lo, j_lo = torch.tril_indices(n_nodes, n_nodes, offset=-1, device=device)
-
-        ei0 = torch.cat([i_up, i_lo])  # Combine them
-        ei1 = torch.cat([j_up, j_lo])
-    else:
-        ei0, ei1 = torch.triu_indices(n_nodes, n_nodes, offset=1, device=device)
-
-    return ei0, ei1
+from gfn.utils.graphs import get_edge_indices
```
Comment (josephdviviano) [inline src/gfn/gym/graph_building.py:14]: cool

File: src/gfn/gym/graph_building.py [hunk 8]
```diff
@@ -89,18 +62,6 @@ def __init__(
             is_directed=is_directed,
         )
 
-    def reset(
```
Comment (josephdviviano) [inline src/gfn/gym/graph_building.py:92]: Why remove this?

File: src/gfn/gym/hypergrid.py [hunk 9]
```diff
@@ -131,17 +131,17 @@ def update_masks(self, states: DiscreteStates) -> None:
...
 
         Args:
             batch_shape: Tuple indicating the shape of the batch.
 
         Returns the batch of random states as tensor of shape (*batch_shape, *state_shape).
         """
-        return torch.randint(
-            0, self.height, batch_shape + self.s0.shape, device=self.device
-        )
+        return torch.randint(0, self.height, batch_shape + self.s0.shape, device=device)
```
Comment (josephdviviano) [inline src/gfn/gym/hypergrid.py:144]: Same comment about it coming from the  env default device if the user does not pass the device.

File: src/gfn/preprocessors.py [hunk 10]
```diff
@@ -31,9 +31,7 @@ def preprocess(self, states: States) -> torch.Tensor:
     def __call__(self, states: States | GraphStates) -> torch.Tensor | GeometricBatch:
         """Transform the states to the input of the neural network, calling the preprocess method."""
         out = self.preprocess(states)
-        if isinstance(out, GeometricBatch):
```
Comment (josephdviviano) [inline src/gfn/preprocessors.py:34]: Why is this being removed? There were cases where a GeometricBatch would be passed to this preprocessor by default. It seems reasonable to keep this as a default.

File: src/gfn/utils/distributions.py [hunk 11]
```diff
@@ -49,51 +47,16 @@ def log_prob(self, sample: torch.Tensor) -> torch.Tensor:
 class GraphActionDistribution(Distribution):
     """A mixture distribution."""
 
-    def __init__(self, logits: TensorDict):
+    def __init__(self, probs: TensorDict):
         """Initializes the mixture distribution.
 
         Args:
-            logits: A TensorDict of logits.
+            probs: A TensorDict of probs.
```
Comment (josephdviviano) [inline src/gfn/utils/distributions.py:54]: Where is the normalization coming from?

Comment (josephdviviano) [reply src/gfn/utils/distributions.py:54]: Oh, I've seen you've moved the logic elsewhere. I'm curious why you would prefer the transformation of logits to probs there, as opposed to here (to me, this seems the logical spot, but I'm open to suggestions).

**Review by josephdviviano — APPROVED** [review]
LGTM! Thanks for the great work. Some comments, but I'm approving so I don't block the eventual merge.

File: src/gfn/actions.py (hunk 1 above)
Comment (hyeok9855) [reply src/gfn/actions.py:49]: Well, for `Actions` (the Tensor action), the device is automatically set the same as `cls.dummy_action` and `cls.exit_action`. We don't have such args for the `GraphActions`; that's why we need it as an argument.

I've changed the default to `None` for the `Actions`.

File: src/gfn/actions.py (hunk 3 above)
Comment (hyeok9855) [reply src/gfn/actions.py:134]: Here, we should ensure that `dummy_actions` and `self.tensor` are on the same device; thus, `self.device` will be fine. Am I misunderstanding your concern?

File: src/gfn/actions.py (hunk 4 above)
Comment (hyeok9855) [reply src/gfn/actions.py:278]: Well, I think I've made the changes needed.

File: src/gfn/gym/box.py (hunk 5 above)
Comment (hyeok9855) [reply src/gfn/gym/box.py:45]: I've also changed the default to `None` here. I'm open to any suggestions!

File: src/gfn/preprocessors.py (hunk 10 above)
Comment (hyeok9855) [reply src/gfn/preprocessors.py:34]: Note first that the `Preprocessor.output_dim` should be equal to `GFNModule.module.input_dim`.

We can't simply assume that the graph action module always sets its `input_dim` to be equal to `out.x.shape[-1]`; **in `GraphEdgeActionMLP`, the `input_dim` should be `n_nodes**2`, not `1 == out.x.shape[-1]`**

File: src/gfn/actions.py (hunk 3 above)
Comment (josephdviviano) [reply src/gfn/actions.py:138]: where is the device from `tensor` set? Is it before we init this class instance?

File: src/gfn/gym/hypergrid.py (hunk 9 above)
Comment (josephdviviano) [reply src/gfn/gym/hypergrid.py:145]: ok thanks

**Comment by josephdviviano** [comment]
thanks for taking the time to answer my questions. i'll merge since I made you wait so long.

File: src/gfn/actions.py (hunk 3 above)
Comment (hyeok9855) [reply src/gfn/actions.py:138]: Yes, it should be.
